### PR TITLE
feat(side-nav): [wip] add mapItem to side-nav

### LINKED ISSUE #1139: Allow adding a prefix to the href attribute of the side-nav
Currently the Link's href is set to item.itemId in the side-nav component. This makes it cumbersome if all Links need to be prefixed.

I can take this up if needed.

### PR DESCRIPTION
**Pull request by chirgjn**
Fixes #1139

#### Description

This allows the user to perform a basic transform on the items array.
Example usecase: add a prefix to each itemId

#### Scope
- [x] Minor: New Feature

### CODE REVIEW

**Comment by gergelyke** [comment]
I wonder, can this be done using the overrides pattern?

**Comment by tajo** [comment]
> I wonder, can this be done using the overrides pattern?

You can't? You could do this transformation (adding a prefix) before you pass the data into the component. However, it's a bit cumbersome since the input data-structure is recursive. It can be handy to have this API and it seems pretty harmless / low maintenance for us.

**Comment by tajo** [comment]
Some test would be great. Otherwise, LGTM.

**Comment by chasestarr** [comment]
it would be nice if this transform function traversed over the options and each option's children, rather than only iterate over the top level

File: src/side-navigation/nav.js [hunk 1]
```diff
@@ -50,14 +52,19 @@ export default class SideNav extends React.Component<NavPropsT> {
       StyledSubNavContainer,
     );
 
-    const renderNavItem = (item: Item, level: number, index) => {
+    const renderNavItem = (item: Item, level: number, index, transformItem) => {
+      if (typeof transformItem === 'function') {
```
[the remark this answers is not in the document]
Comment (chasestarr) [reply src/side-navigation/nav.js]: do you notice it performing poorly? I wouldn't worry until we have a use-case where the perf improvement is needed

[the remark this answers is not in the document]
Comment (chirgjn) [reply src/side-navigation/nav.js]: I'm not that worried about performance. My main concern was repeated mutation. Suppose add a suffix to a title using transformItem. If i don't use flow then I have no way of knowing that transformItem should be idempotent.

[the remark this answers is not in the document]
Comment (chasestarr) [reply src/side-navigation/nav.js]: you mean, mutating the 'source' data when running the transform fn? that's often out of the norm, and should be visibly apparent when testing the UI. maybe calling the prop `mapItem` would be more descriptive?

[the remark this answers is not in the document]
Comment (chirgjn) [reply src/side-navigation/nav.js]: +1 `mapItem` feels better

File: src/side-navigation/types.js [hunk 2]
```diff
@@ -40,12 +40,14 @@ type renderItemT = (
   },
 ) => React.Node;
 
+type mapItemT = (item: Item) => Item;
```
Comment (chasestarr) [inline src/side-navigation/types.js]: please in-line on L68 if it's not used anywhere else

Comment (chirgjn) [reply src/side-navigation/types.js]: Done :)

**Comment by chirgjn** [comment]
Do I need to add it to the docs?

**Comment by chasestarr** [comment]
it will be automatically added to the api docs. check out the preview here: https://deploy-preview-1173--baseui.netlify.com/components/sidenav/#side-navigation-api

**Comment by chirgjn** [comment]
@chasestarr I see on of the CI tests are failing, do I need to make any changes to get this merged?

**Comment by chasestarr** [comment]
Sorry - I'll get this merged in today

**Comment by chasestarr** [comment]
Thank You!

**Pull request merged**